diff --git a/gvsoc/gvsoc/models/cpu/iss/flexfloat/flexfloat.c b/gvsoc/gvsoc/models/cpu/iss/flexfloat/flexfloat.c
index 215f53a7..ee15e1e2 100644
--- a/gvsoc/gvsoc/models/cpu/iss/flexfloat/flexfloat.c
+++ b/gvsoc/gvsoc/models/cpu/iss/flexfloat/flexfloat.c
@@ -381,7 +381,7 @@ INLINE void ff_init_longdouble(flexfloat_t *obj, long double value, flexfloat_de
     flexfloat_sanitize(obj);
 }
 
-INLINE void ff_init_float128(flexfloat_t *obj, __float128 value, flexfloat_desc_t desc) {
+INLINE void ff_init_float128(flexfloat_t *obj, _Float128 value, flexfloat_desc_t desc) {
     obj->value = (fp_t)value;
     #ifdef FLEXFLOAT_TRACKING
     obj->exact_value = (fp_t)value;;
@@ -447,8 +447,8 @@ INLINE long double ff_get_longdouble(const flexfloat_t *obj) {
     return (long double)(*((const fp_t *)(&(obj->value))));
 }
 
-INLINE __float128 ff_get_float128(const flexfloat_t *obj) {
-    return (__float128)(*((const fp_t *)(&(obj->value))));
+INLINE _Float128 ff_get_float128(const flexfloat_t *obj) {
+    return (_Float128)(*((const fp_t *)(&(obj->value))));
 }
 
 
diff --git a/gvsoc/gvsoc/models/cpu/iss/flexfloat/flexfloat.h b/gvsoc/gvsoc/models/cpu/iss/flexfloat/flexfloat.h
index 0ab4f47f..6b0e5b8f 100644
--- a/gvsoc/gvsoc/models/cpu/iss/flexfloat/flexfloat.h
+++ b/gvsoc/gvsoc/models/cpu/iss/flexfloat/flexfloat.h
@@ -112,7 +112,7 @@ static uint64_t smallest_norm_neg_value[2] = {0x0000000000000000, 0x800100000000
 #define NUM_BITS_FRAC (112)
 typedef int128_t int_t;
 typedef uint128_t uint_t;
-typedef __float128 fp_t;
+typedef _Float128 fp_t;
 #define PRINTF_FORMAT "%.38Lf"
 #endif /* FLEXFLOAT_ON_QUAD */
 
@@ -207,7 +207,7 @@ void ff_init(flexfloat_t *obj, flexfloat_desc_t desc);
 void ff_init_float(flexfloat_t *obj, float value, flexfloat_desc_t desc);
 void ff_init_double(flexfloat_t *obj, double value, flexfloat_desc_t desc);
 void ff_init_longdouble(flexfloat_t *obj, long double value, flexfloat_desc_t desc);
-void ff_init_float128(flexfloat_t *obj, __float128 value, flexfloat_desc_t desc);
+void ff_init_float128(flexfloat_t *obj, _Float128 value, flexfloat_desc_t desc);
 void ff_init_int(flexfloat_t *obj, int value, flexfloat_desc_t desc);
 void ff_init_long(flexfloat_t *obj, long value, flexfloat_desc_t desc);
 void ff_cast(flexfloat_t *obj, const flexfloat_t *source, flexfloat_desc_t desc);
@@ -217,7 +217,7 @@ void ff_cast(flexfloat_t *obj, const flexfloat_t *source, flexfloat_desc_t desc)
 float ff_get_float(const flexfloat_t *obj);
 double ff_get_double(const flexfloat_t *obj);
 long double ff_get_longdouble(const flexfloat_t *obj);
-__float128 ff_get_float128(const flexfloat_t *obj);
+_Float128 ff_get_float128(const flexfloat_t *obj);
 
 
 // Artihmetic operators
